File: packages/supervisor-client-ts/src/jwt.ts
function b64urlEncode(bytes: Uint8Array): string {
  let s = "";
  for (const b of bytes) s += String.fromCharCode(b);
  return btoa(s).replace(/=+$/, "").replace(/\+/g, "-").replace(/\//g, "_");
}

function str2bytes(s: string): Uint8Array {
  return new TextEncoder().encode(s);
}

async function hmacSha256(key: string, msg: string): Promise<Uint8Array> {
  const k = await crypto.subtle.importKey(
    "raw",
    str2bytes(key),
    { name: "HMAC", hash: "SHA-256" },
    false,
    ["sign"],
  );
  const sig = await crypto.subtle.sign("HMAC", k, str2bytes(msg));
  return new Uint8Array(sig);
}

export async function signHS256(claims: Record<string, unknown>, secret: string): Promise<string> {
  const header = { alg: "HS256", typ: "JWT" };
  const h = b64urlEncode(str2bytes(JSON.stringify(header)));
  const p = b64urlEncode(str2bytes(JSON.stringify(claims)));
  const sig = await hmacSha256(secret, `${h}.${p}`);
  return `${h}.${p}.${b64urlEncode(sig)}`;
}

export async function buildToken(
  appId: string,
  scopes: string[],
  secret: string,
  ttlSeconds = 300,
): Promise<string> {
  const now = Math.floor(Date.now() / 1000);
  return signHS256({ sub: appId, scopes, iat: now, exp: now + ttlSeconds }, secret);
}

File: packages/supervisor-client-py/src/supervisor_client/_jwt.py
from __future__ import annotations

import base64
import hashlib
import hmac
import json
from datetime import datetime, timedelta, timezone

# `datetime.UTC` is Python 3.11+. Alias for 3.10 compatibility so the
# client can run from user repos on older Pythons (Clio is on 3.10).
UTC = timezone.utc
from typing import Any


def _b64url(raw: bytes) -> str:
    return base64.urlsafe_b64encode(raw).rstrip(b"=").decode()


def sign_hs256(claims: dict[str, Any], secret: str) -> str:
    header = {"alg": "HS256", "typ": "JWT"}
    h = _b64url(json.dumps(header, separators=(",", ":")).encode())
    p = _b64url(json.dumps(claims, separators=(",", ":"), default=str).encode())
    sig = hmac.new(secret.encode(), f"{h}.{p}".encode(), hashlib.sha256).digest()
    return f"{h}.{p}.{_b64url(sig)}"


def build_token(app_id: str, scopes: list[str], secret: str, ttl_seconds: int = 300) -> str:
    now = datetime.now(UTC)
    return sign_hs256(
        {
            "sub": app_id,
            "scopes": scopes,
            "iat": int(now.timestamp()),
            "exp": int((now + timedelta(seconds=ttl_seconds)).timestamp()),
        },
        secret,
    )
